### PR TITLE
Added first working FMCIB model container

### PR DESCRIPTION
**Pull request by surajpaib**

### CODE REVIEW

File: models/fmcib/dockerfiles/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+RUN wget https://zenodo.org/records/10528450/files/model_weights.torch?download=1 -O /app/model_weights.torch
+
+RUN mkdir models
+RUN mkdir models/fmcib
+
+# Install FMCIB package, should install everything else ...
+RUN pip install foundation-cancer-image-biomarker --pre
+
+ENTRYPOINT ["python3", "-m", "mhubio.run"]
```
Comment (LennyN95) [inline models/fmcib/dockerfiles/Dockerfile]: We can use `mhub.run` instead ;)

File: models/fmcib/utils/FMCIBRunner.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+from fmcib.models import fmcib_model 
+import SimpleITK as sitk
+from mhubio.core import Instance, InstanceData, IO, Module
+from fmcib.preprocessing import preprocess
+
+
+class FMCIBRunner(Module):
+    @IO.Instance()
+    @IO.Input('in_data', 'nrrd:mod=ct', the='Input NRRD file')
+    @IO.Input('in_mask', 'nrrd|json', the='Tumor mask for the input NRRD file')
```
Comment (LennyN95) [inline models/fmcib/utils/FMCIBRunner.py]: In the code, we need to check which data type was provided. 
As mentioned, I suggest we outsource the centroid extraction into a module (I can draft one) to make it all more spread, verbose, and transparent (this is more of a nice-to-have than a technical requirement but I think we agreed on this :D).

File: models/fmcib/utils/FMCIBRunner.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    @IO.Output('feature_json', 'features.json', "json", bundle='model', the='output JSON file')
```
Comment (LennyN95) [inline models/fmcib/utils/FMCIBRunner.py]: If we accept json as an input, I suggest we create a schema file and validate it properly. Then, why not have a schema for the output json, too? What do you think? Should be effortless to do but makes makes it crystal clear what to expect from the model output!

File: models/fmcib/utils/FMCIBRunner.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+        image = preprocess(input_dict)
+        image = image.unsqueeze(0)
+        model = fmcib_model()
+
+        model.eval()
+        with torch.no_grad():
+            features = model(image)
+
+        feature_dict = {f"feature_{idx}": feature for idx, feature in enumerate(features.flatten().tolist())}
```
Comment (LennyN95) [inline models/fmcib/utils/FMCIBRunner.py]: Our [logical data wrappers ](https://github.com/MHubAI/documentation/blob/main/documentation/mhubio/how_to_write_an_mhubio_module.md#handling-logical-output-data) could be a good fit here. 

Although we can (and should) return the original output, this, besides adding verbosity, allows us to bring the output into every format using e.g. our ReportExporter right from the config.

**Review by LennyN95 — COMMENTED** [review]
Looking good so far! Some comments, let's chat per zoom if anything is unclear or could be improved ;)

File: models/fmcib_radiomics/dockerfiles/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+LABEL authors="bspai@bwh.harvard.edu"
+
+ARG MHUB_MODELS_REPO
+# Add pull models repo command here after local testingRUN
+RUN buildutils/import_mhub_model.sh fmcib_radiomics ${MHUB_MODELS_REPO}
+RUN wget https://zenodo.org/records/10528450/files/model_weights.torch?download=1 -O /app/model_weights.torch
+
+
+RUN mkdir models
+RUN mkdir models/fmcib
```
Comment (LennyN95) [inline models/fmcib_radiomics/dockerfiles/Dockerfile]: Not sure if we need that?
And if we do `RUN mkdir -p models/fcib`.

File: models/fmcib_radiomics/dockerfiles/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,20 @@
+FROM mhubai/base:latest
+
+LABEL authors="bspai@bwh.harvard.edu"
+
+ARG MHUB_MODELS_REPO
+# Add pull models repo command here after local testingRUN
+RUN buildutils/import_mhub_model.sh fmcib_radiomics ${MHUB_MODELS_REPO}
+RUN wget https://zenodo.org/records/10528450/files/model_weights.torch?download=1 -O /app/model_weights.torch
```
Comment (LennyN95) [inline models/fmcib_radiomics/dockerfiles/Dockerfile]: Maybe we choose a distinct file name or place the weights inside a folder to keep us the option to add then later successive models and have all their weights organized. Let's spend a minute thinking about this.

File: models/fmcib_radiomics/meta.json [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        "description": "CT imaging data containing lesions of interest, such as nodules or tumors",
+        "format": "DICOM",
+        "modality": "CT",
+        "slicethickness": "5mm",
+        "bodypartexamined": "Whole",
+        "non-contrast": true,
+        "contrast": true
+      },
+      {
+        "label": "Center of mass",
```
Comment (LennyN95) [inline models/fmcib_radiomics/meta.json:19]: This doesn't match with the current default workflow but we can address this later and decide which of the two (json / mask) we want to keep as the default. I'd choose whatever is the easiest and most standard one.

File: models/fmcib_radiomics/utils/FMCIBRunner.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        x, y, z = centroid
+
+        input_dict = {
+            "image_path": in_data.abspath,
+            "coordX": x,
+            "coordY": y,
+            "coordZ": z,
+        }
+
+        image = preprocess(input_dict)
```
Comment (LennyN95) [inline models/fmcib_radiomics/utils/FMCIBRunner.py]: I was wondering, don't you have a direct entrypoint (such that we don't need to import torch and do the channel arrangement manually?).

Something like
```Python
feature_dict = fmcib(input_dict)
```

File: models/fmcib_radiomics/meta.json (hunk 7 above)
Comment (LennyN95) [reply models/fmcib_radiomics/meta.json:19]: I provided an alternative workflow that can start from the JSON file. I've some ideas on how we could provide a workflow starting from Dicom (which should be the default then) that I want to discuss later!

**Comment by surajpaib** [comment]
I tested the default workflow, made some small fixes and it all looks good. 

Also tested the slicer workflow on a folder structure that looks like this
image

Using the `PatientID` folder as the lead for the structure. It all works very well 🎆 

For the testing, 
This should pull all the DICOM data: `s3://idc-open-data-cr/40db0070-9773-413a-8d7e-c91c37c0006b/*` and I'll make a release with the centroids.json and the features.json

File: models/fmcib_radiomics/config/from_slicer.yml
Comment (surajpaib) [inline models/fmcib_radiomics/config/from_slicer.yml:1]: This is here for now just to show you what I did with the slicer stuff, we can just restructure this later as you suggest

**Comment by surajpaib** [comment]
Hmm, I'm not too sure how you'd want the format to be for these on the release (since we don't have any DICOM + json workflows yet and I am not too sure how to do that myself without spending a lot of time 😓 ) 

So I've attached both the features.json and the centroids.json for the IDC DICOM scan that I linked above. Maybe you have an idea of how to do this best?

[features.json](https://github.com/MHubAI/models/files/14520685/features.json)
[centroids.json](https://github.com/MHubAI/models/files/14520712/centroids.json)

**Comment by LennyN95** [comment]
That's great! I'll implement the dicom workflow later today!

**Comment by surajpaib** [comment]
@LennyN95 Here is another sample. 

Image: `s3://idc-open-data-cr/18b83380-0f27-42f2-af7c-65e2d23668ac/*`
Centroid: 
[LUNG1-076_centroids.json](https://github.com/MHubAI/models/files/14528031/LUNG1-076_centroids.json)
Features: 
[LUNG1-076_features.json](https://github.com/MHubAI/models/files/14528037/LUNG1-076_features.json)

**Comment by LennyN95** [comment]
/test

```yaml
sample:
  idc_version: 17.0
  data:
  - SeriesInstanceUID: 1.3.6.1.4.1.32722.99.99.298991776521342375010861296712563382046
    aws_url: s3://idc-open-data-cr/40db0070-9773-413a-8d7e-c91c37c0006b/*
    path: dicom
  - url: https://github.com/MHubAI/models/files/14520712/centroids.json
    path: 1.3.6.1.4.1.32722.99.99.298991776521342375010861296712563382046.json

  - SeriesInstanceUID: 1.3.6.1.4.1.32722.99.99.309695689942006699082558764031786785731
    aws_url: s3://idc-open-data-cr/18b83380-0f27-42f2-af7c-65e2d23668ac/*
    path: dicom
  - url: https://github.com/MHubAI/models/files/14528031/LUNG1-076_centroids.json
    path: 1.3.6.1.4.1.32722.99.99.309695689942006699082558764031786785731.json

reference:
  url: https://github.com/MHubAI/models/files/14612917/fmcib_mhub_default_output.zip
```

---

<details><summary>Test Results (24.03.15_11.57.01_d91gDb1QPt)</summary>

```yaml
id: dbce25f8-eaa8-411f-a4ec-d56e6e9f7df6
date: '2024-03-15 11:57:56'
checked_files:
- file: features.json
  path: /app/test/src/1.3.6.1.4.1.32722.99.99.309695689942006699082558764031786785731/features.json
  checks:
  - checker: DataFileCheck
    notes:
    - label: Value Match
      description: These keys have identical values
      info: feature_0,feature_1,feature_2,feature_3,feature_4,feature_5,feature_6,feature_7,feature_8,feature_9,feature_10,feature_11,feature_12,feature_13,feature_14,feature_15,feature_16,feature_17,feature_18,feature_19
        (+ 4076 more)
  - checker: SizeCheck
- file: features.json
  path: /app/test/src/1.3.6.1.4.1.32722.99.99.298991776521342375010861296712563382046/features.json
  checks:
  - checker: DataFileCheck
    notes:
    - label: Value Match
      description: These keys have identical values
      info: feature_0,feature_1,feature_2,feature_3,feature_4,feature_5,feature_6,feature_7,feature_8,feature_9,feature_10,feature_11,feature_12,feature_13,feature_14,feature_15,feature_16,feature_17,feature_18,feature_19
        (+ 4076 more)
  - checker: SizeCheck
summary:
  files_missing: 0
  files_extra: 0
  checks:
    DataFileCheck:
      files: 2
    SizeCheck:
      files: 2
conclusion: true
```

</details>
